<launch>

<arg name="driver_decimation" default="false"/>

<arg name="gazebo" default="true"/>

<arg name="debug" default="false"/>

<arg name="debug_prefix" value="" unless="$(arg debug)"/>
<arg name="debug_prefix" value="xterm -e gdb --args" if="$(arg debug)"/>
<arg name="debug_prefix1" value="valgrind --tool=memcheck " if="$(arg debug)"/>

<arg name="depth_image_raw" default="/camera/depth/image_raw"/>
<arg name="depth_info_raw" default="/camera/depth/camera_info"/>

<arg name="depth_image_dec" default="/camera/depth_decimated/image"/>
<arg name="depth_info_dec" default="/camera/depth_decimated/camera_info"/>

<!--If driver_decimation is on, then the temporary topic is set to the raw stream. Otherwise, it is set to the decimated stream-->
<arg name="depth_image_topic_t" value="$(arg depth_image_raw)" if="$(arg driver_decimation)"/>
<arg name="depth_info_topic_t" value="$(arg depth_info_raw)" if="$(arg driver_decimation)"/>
<arg name="depth_image_topic_t" value="$(arg depth_image_dec)" unless="$(arg driver_decimation)"/>
<arg name="depth_info_topic_t" value="$(arg depth_info_dec)" unless="$(arg driver_decimation)"/>

<!--If gazebo is enabled, we use the raw stream. Otherwise, we use whatever was selected above-->
<arg name="depth_image_topic" value="$(arg depth_image_raw)" if="$(arg gazebo)"/>
<arg name="depth_info_topic" value="$(arg depth_info_raw)" if="$(arg gazebo)"/>
<arg name="depth_image_topic" value="$(arg depth_image_topic_t)" unless="$(arg gazebo)"/>
<arg name="depth_info_topic" value="$(arg depth_info_topic_t)" unless="$(arg gazebo)"/>

<!--The selected topics above are used, though they can still be remapped from the command line-->
<arg name="depth_image" default="$(arg depth_image_topic)"/>
<arg name="depth_info" default="$(arg depth_info_topic)"/>

<arg name="camera" default="camera"/>

<arg name="parfor" default="true"/>

<arg name="nodelet" default="true" unless="$(arg gazebo)"/>
<arg name="nodelet" default="false" if="$(arg gazebo)"/>

<arg name="spinner" default="false"/>

<arg name="depth_res" default="8"/>
<!-- depth_res options:
  2: VGA    640x480
  5: QVGA   320x240
  8: QQVGA  160x120
-->

<!--If driver decimation enabled, set depth camera resolution to lowest-->
<group if="$(arg driver_decimation)">

  <param name="/$(arg camera)/driver/depth_mode" value="$(arg depth_res)" />
  <node name="dynparam_set_depth_resolution" pkg="dynamic_reconfigure" type="dynparam" args="set /$(arg camera)/driver depth_mode $(arg depth_res)"/> 
</group>

<group unless="$(arg driver_decimation)">

</group>

<param name="enable_parallel_loop" value="$(arg parfor)"/>

<arg name="controller_name" value="pips_controller_nodelet" if="$(arg nodelet)"/>
<arg name="controller_name" value="pips_controller_node" unless="$(arg nodelet)"/>



<!-- other topics to remap: enable, disable, GenAndTest/desired_path -->
<node name="$(arg controller_name)" pkg="nodelet" type="nodelet" args="load pips_trajectory_testing/pips_controller_nodelet camera_nodelet_manager" ns="$(arg camera)" if="$(arg nodelet)" >
    <remap from="camera/depth/image_raw" to="depth_decimated/image" />
    <remap from="camera/depth/camera_info" to="depth_decimated/camera_info" />
    <remap from="GenAndTest/tested_paths" to="/GenAndTest/tested_paths"/>
    <remap from="GenAndTest/collision_points" to="/GenAndTest/collision_points"/>
    <remap from="odom" to="/odom"/>
    <remap from="cmd_vel_mux/input/navi" to="/cmd_vel_mux/input/navi"/>
    <remap from="desired_trajectory" to="/desired_trajectory"/>
    <remap from="desired_odom" to="/desired_odom"/>
    <remap from="transformed_trajectory" to="/transformed_trajectory"/>
    <param name="odom_spinner" value="$(arg spinner)"/>
</node>




<group unless="$(arg gazebo)">
  <node name="dynparam_pips_controller" pkg="dynamic_reconfigure" type="dynparam" args="load $(arg controller_name) $(find pips_trajectory_testing)/param/PipsController.yaml"/> 
  
  <node name="dynparam_genandtest" pkg="dynamic_reconfigure" type="dynparam" args="load $(arg controller_name)/GenAndTest $(find pips_trajectory_testing)/param/GenAndTest.yaml"/> 
  
  
</group>



<node name="$(arg controller_name)" pkg="pips_trajectory_testing" type="pips_controller_node" output="screen" args="" launch-prefix="$(arg debug_prefix)" unless="$(arg nodelet)">
    <remap from="/camera/camera/depth/image_raw" to="$(arg depth_image)" />
    <remap from="/camera/camera/depth/camera_info" to="$(arg depth_info)" />
    <param name="odom_spinner" value="$(arg spinner)"/>
</node>


  
</launch>
